docs/content/utilities/batch.mdx

```diff
@@ -0,0 +1,95 @@
+---
+title: Batch
+description: Utility
+---
+
+import Note from "../../src/components/Note"
+
+The batch utility provides an abstraction to process a batch event. Useful for lambda integrations with [AWS SQS](https://aws.amazon.com/sqs/), [AWS Kinesis](https://aws.amazon.com/kinesis/) and [AWS DynamoDB Streams](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.html).
+It also provides base classes (`BaseProcessor`, `BasePartialProcessor`) allowing you to create your **own** batch processor.
+
+**Key Features**
+
+* Run batch processing logic with a clean interface;
+* Middleware and context to handle a batch event;
+* Removal of successful messages for [AWS SQS](https://aws.amazon.com/sqs/) batch - in case of partial failure.
+
+**IAM Permissions**
+
+This utility requires additional permissions to work as expected. See the following table:
+
+Processor | Function/Method | IAM Permission
+|---------|-----------------|---------------|
+PartialSQSProcessor | `_clean` | `sqs:DeleteMessageBatch`
+
+### PartialSQSProcessor
+
+A special batch processor which aims to `clean` your SQS:Queue if one or more (not all) records of the batch fails.
+A batch's partial failure sends back all the records to the queue, reprocessing this batch until all records succed.
+This processor exists to improve performance in such cases, deleting successful messages of a batch with partial failure.
+
+### Middleware
+
+```python:title=app.py
+from aws_lambda_powertools.utilities.batch import batch_processor, PartialSQSProcessor
+
+def record_handler(record):
+    return record["body"]
+
+# highlight-start
+@batch_processor(record_handler=record_handler, processor=PartialSQSProcessor())
+# highlight-end
+def lambda_handler(event, context):
+    return {"statusCode": 200}
+```
+
+## Create your own processor
+
+You can create your own batch processor by inheriting the `BaseProcessor` class, and implementing `_prepare()`, `_clean` and `_process_record()`.
+It's also possible to inherit the `BasePartialProcessor` which contains additional logic to handle a partial failure and keep track of record status.
+
+Here is an example implementation of a DynamoDBStream custom processor:
+
+```python:title=custom_processor.py
+import json
+
+from aws_lambda_powertools.utilities.batch import BaseProcessor, batch_processor
+import boto3
+
+class DynamoDBProcessor(BaseProcessor):
+
+    def __init__(self, queue_url: str):
+        self.queue_url = queue_url
+        self.client = boto3.client("sqs")
```

Comment:
I think you copied the wrong class the sample is about dynamodb this is sqs 😉

Reply:
Nope ! Improved the example with some comments. Can you take a look again ? Case it's unclear I'll switch the example.

Reply:
Right, now I get the intention would `DDBStreamToSQSProcessor` be a better name then? Or something among those lines `DDBStreamProcessor` is very generic and the sample has some specific SQS behaviour in there.

Reply:
Thanks for working on the docs @gmcrocetti, much appreciated ;) It's clearer now, and it helped me confirm my initial concerns with it:

Shouldn't this utility be primarily focused on SQS Partial batch failures? 

I understand the appeal for having an interface to handle batch processing, though it makes more sense in non-Lambda function's environment. If customers have to implement their own batch processors, then this means we (as a Provider) should step back and work on better constructs to allow customers to have less code since code is liability - For example, DynamoDB, Kinesis, S3 Batch Operations all benefit from async controls, max retries, etc. requiring zero code to be maintained.

Right now, my vote would be for:

* [ ] Have only `BasePartialProcessor` so that customers could opt-in for your great SQS partial implementation, or create their own
* [ ] Remove any non-partial, non-SQS code unless we have a demand for it (80% tenet)
* [ ] Reword the custom processor doc section to focus on a custom SQS partial failure processor, where a customer could add additional logging or tracing annotations, etc.

cc @cakepietoast for a second opinion as I'd really like to include this in our next release this week, but am conflicted on the value of creating a custom processor for non-partial, non-SQS processing vs maintaining it.

Once again @gmcrocetti, I **really appreciate all your efforts, and sorry this is taking longer than you might have expected** - We want to make sure we strike a balance on customers knowing how easy it is to see the value prop of your utility and use it vs maintenance and customer demand.

Reply:
@heitorlessa I agree. This is an awesome PR with "just" the functionality to gracefully handle partial failures in SQS. I think its best that we focus on the pain point that was raised in the RFC, which is very specific to SQS. I'm open to implementing other stream processing functionality in the future - perhaps as a separate utility - but we need to give that some more thought beforehand.

Reply:
Hehehhe, thanks for worrying so much, @heitorlessa , no problem at all. Sent a new commit ! :)